### PR TITLE
Adding Python Expression documentation

### PR DESCRIPTION
**Pull request by parmitam**
Addition to the documentation to add python expression documentation

### CODE REVIEW

File: docs/myria-python/index.md [hunk 1]
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+    import math
+    n = dt[0][0]
+    if n % 2 == 0 and n > 2: 
+        return False
+    for i in range(3, int(math.sqrt(n)) + 1, 2):
+        if n % i == 0:
+            return 0
+    return 1
+
+MyriaPythonFunction(pyIsPrime, outType).register()
```
Comment (BrandonHaynes) [inline docs/myria-python/index.md]: Why not a decorator?  Users shouldn't ever need to call `register` explicitly.

Comment (parmitam) [reply docs/myria-python/index.md]: I haven't got the decorator to work, feel free to add it when you have it working correctly.thanks

Comment (BrandonHaynes) [reply docs/myria-python/index.md]: Is there a GitHub issue with the problem you're encountering?  I know that @orzikhd has been able to use them.  Thanks.

File: docs/myria-python/index.md [hunk 2]
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
 print query.status
 ```
 
+### Part 4. Python Expressions
+
+Myria allows arbitrary Python Functions as expressions which can be used in transformations like map, flatmap and aggregates.
+#### Registering a Python Function
+Python function needs to be registered before it can be used. Myria-python can be used to register python functions, list registered python functions as well as get details of the registered function.
```
Comment (jortiz16) [inline docs/myria-python/index.md]: Maybe reword this as... "Myria-Python can be used to register new Python functions, list existing registered Python functions, and to retrieve details about existing registered functions"

Also I did not see an example within this block where you list how to get details about the registered functions. You only show how to list them.

File: docs/myria-python/index.md [hunk 3]
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+            return 0
+    return 1
+
+MyriaPythonFunction(pyIsPrime, outType).register()
+
+#List registered functions
+connection.get_functions()
+```
+#### Invoking Python Expression in MyriaL
+A registered Python function can then be used in a MyriaL  query as an operator as:
```
Comment (jortiz16) [inline docs/myria-python/index.md]: "A registered Python function can be used in a MyriaL query as an operator:"

Is this an operator though? Isn't it an expression?

File: docs/myria-python/index.md [hunk 4]
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+T1 = scan(TwitterK);
+isPrime = [from T1 emit pyIsPrime(T1.src) as isPrime, T1.src, T1.dst];
+store( isPrime, TwitterK_isPrime);
+""", connection=connection)
+q.status
+
+```
+#### Python Expressions in UDAs
+Python functions call also be used in User Defined Aggregates in MyriaL.
+
```
Comment (jortiz16) [inline docs/myria-python/index.md]: Missing "can"

File: docs/myria-python/index.md [hunk 5]
```diff
@@ -178,6 +178,89 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+    --emit
+    [tm];
+};
+t = scan(public:adhoc:raw);
+results = [from t emit t.subjid, t.imgid, foo(t.subjid, t.imgid,t.img) as vox];
+STORE(results, results);
+""", connection=connection)
+
+q.status
+```
```
Comment (jortiz16) [inline docs/myria-python/index.md]: Also, how can we run these? I wanted to check if all the examples work. Do I need to use a separate branch for anything via Myria-Cluster? Doesn't seem like the Myria Demo cluster can run these queries....

**Review by jortiz16 — CHANGES_REQUESTED** [review]

File: docs/myria-python/index.md (hunk 3 above)
Comment (parmitam) [reply docs/myria-python/index.md]: it is an expression, but it can be used as an operator, i'll simplify the language, i can see why this is confusing.

File: docs/myria-python/index.md (hunk 5 above)
Comment (parmitam) [reply docs/myria-python/index.md]: you need to deploy a cluster with  --unprovisioned.
@senderista  when will we have updated images for this?

File: docs/myria-python/index.md (hunk 1 above)
Comment (orzikhd) [reply docs/myria-python/index.md]: I just tried it out and was able to use the @myria_function decorator, but it should be noted that I'm on a slightly old version of Myria at this point. The following code worked:
```
from myria import *
%load_ext myria
connection = MyriaConnection(rest_url='http://localhost:8753')
%connect http://localhost:8753
from myria.udf import myria_function

@myria_function(name='bar', output_type="LONG_TYPE", connection=connection)
def bar(dt):
    import math
    n = dt[0][0]
    if n % 2 == 0 and n > 2: 
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return 0
    return 1
```

Comment (parmitam) [reply docs/myria-python/index.md]: I was hoping just 
```python
@myria_function(output_type="LONG_TYPE")
def bar(dt):
    import math
    n = dt[0][0]
    if n % 2 == 0 and n > 2: 
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return 0
    return 1
```
 would work

Comment (parmitam) [reply docs/myria-python/index.md]: Also, @BrandonHaynes  could we please update the demo cluster so these work on the demo cluster. 
a cluster deployed with myria-cluster and --unprovisioned flag works, so I am not sure what needs to happen.  
I would not be  comfortable pushing the doc changes if they don't work on the demo cluster. Thanks

File: docs/myria-python/index.md (hunk 2 above)
Comment (parmitam) [reply docs/myria-python/index.md]: done.

File: docs/myria-python/index.md (hunk 3 above)
Comment (parmitam) [reply docs/myria-python/index.md]: Done

**Review by jortiz16 — APPROVED** [review]

File: docs/myria-python/index.md [hunk 6]
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+Python functions can also be used in User Defined Aggregates in MyriaL.
+
+```python
+#register a UDA
+def udfSum(dt):
+    import numpy as np
+    tuplist = dt
+    state = None
+    for i in tuplist:
+        imgid = i12]
```
Comment (jortiz16) [inline docs/myria-python/index.md]: There is a syntax error here. After `i12` there shouldn't be a bracket

File: docs/myria-python/index.md [hunk 7]
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+        img = np.asarray(i[2])
+        shape = img.shape + (5,)
+        if state is None:
+            state = np.empty(shape)
+            state[:,:,:,imgid] = img
+        else:
+            state[:,:,:,imgid] = img
+    return (state)
+
+MyriaPythonFunction(udfAgg, outType).register()
```
Comment (jortiz16) [inline docs/myria-python/index.md]: This should be`udfSum`

File: docs/myria-python/index.md [hunk 8]
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+store( isPrime, TwitterK_isPrime);
+""", connection=connection)
+q.status
+
+```
+#### Python Expressions in UDAs
+Python functions can also be used in User Defined Aggregates in MyriaL.
+
+```python
+#register a UDA
```
Comment (jortiz16) [inline docs/myria-python/index.md]: I think this should be `UDF`

File: docs/myria-python/index.md (hunk 6 above)
Comment (parmitam) [reply docs/myria-python/index.md]: fixed.

File: docs/myria-python/index.md (hunk 1 above)
Comment (BrandonHaynes) [reply docs/myria-python/index.md]: If you set a default connection (`MyriaRelation.DefaultConnection = c`) then you don't have to specify it when registering.  If you haven't, myria-python has no idea which (and there may be >1) to use when connecting.

File: docs/myria-python/index.md [hunk 9]
```diff
@@ -178,6 +178,98 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+            return 0
+    return 1
+
+MyriaPythonFunction(pyIsPrime, outType).register()
+
+#List registered functions
+connection.get_functions()
+
+#List details of a registered function
+connection.get_function('pyIsPrime')
```
Comment (BrandonHaynes) [inline docs/myria-python/index.md]: myria-python has a higher-level API for getting functions (as entities, not JSON documents) as `MyriaFunction.get(name, connection)` and `MyriaFunction.get_all(connection)`.  There should be no for users to ever interact directly with the MyriaConnection facade.

I'll clean this up after I update the demo cluster.

File: docs/myria-python/index.md [hunk 10]
```diff
@@ -178,6 +178,95 @@ query = MyriaQuery.parallel_import(relation=relation, work=work)
...
+q = MyriaQuery.submit("""
+  uda aggregate(subjid,imgid, img) {
+    --init
+    [ b'' as tm];
+    --update
+    [udfSum(subjid ,imgid, img )];
+    --emit
+    [tm];
+  };
+  t = scan(public:adhoc:raw);
```
Comment (BrandonHaynes) [inline docs/myria-python/index.md]: This relation doesn't exist on a vanilla Myria install (or on demo) -- need to stage the content on S3 and do an explicit load so that the query works.

**Comment by BrandonHaynes** [comment]
UDF documentation is corrected.  I can remove two additional explicit connection references when https://github.com/uwescience/myria-python/pull/73 is merged.

The relation `public:adhoc:raw` needs to be explicitly loaded from a stable external source so that this works on a vanilla Myria install.

File: docs/myria-python/index.md (hunk 5 above)
Comment (senderista) [reply docs/myria-python/index.md]: just for the record, we have updated images that can run the examples.

**Comment by senderista** [comment]
Is this PR only blocking on our locating the data for the relation `public:adhoc:raw`? @parmitam can you address this?

**Comment by parmitam** [comment]
Made a change to the UDA so we don't need the image relation, it works on twitterK relation, the UDA is trivial, but it gets the point across.
